### ISSUE TITLE
Validate that characters are part of the PAE character set

### ISSUE THREAD

**Issue by ahankinson** [rism-digital/verovio#2510]
When entering PAE input, Verovio should check whether the characters entered are even part of the PAE character set, and should trigger a warning if characters are entered that do not belong at all.

Related to rism-digital/muscat#901

**Comment by craigsapp** [comment]
Verovio does not validate MEI or MusicXML files... 😜 

But validating characters should be easy for PAE if  only UTF-7 characters are allowed, by creating a static array of 256 elements with true/false state for each character represented by that position:

```cpp
vector<char> validPAE {
     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
     0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
etc. a total of 256 times
};
```

Then set each position in the array to `1` if the ASCII character is allowed.   And then check each position in the array, as long as the character is less than 256 (unsigned char) or greater than 0 for (signed char):

```cpp
   for (char &ch : str) {
      if (ch < 0) {
         cerr << "Invalid PAE byte: " << ch << endl;
         continue;
      }
      if (!validPAE[ch]) {
         cerr << "Invalid PAE character: " << ch << endl;
      }
   }
```

**Comment by ahankinson** [comment]
> Verovio does not validate MEI or MusicXML files

Are you volunteering? :-D

Is your method much better than just creating a list of characters and validating whether a character is in that list? The list of characters is [pretty small](https://github.com/rism-digital/paec-grammar/blob/main/paec.g4#L332-L380), so it would indeed be a very sparse array. And debugging would be hard.

**Comment by craigsapp** [comment]
> Are you volunteering? :-D

If no one beats me to it, I could :-)

> Is your method much better than just creating a list of characters and validating whether a character is in that list?

Looking at the valid character sequence, there are gaps, so it would be more efficient to do in the method that I described since you cannot take advantage of character ranges very much, so your algorithm is `O(M*N)` and mine is `O(N)`.  So it will be saving a microsecond or so, given the length of PAE code, and that would save you 5 seconds over 5 million incipits :-).  For debugging, you just have to look at the following code for errors and regenerate the array if the list is incorrect, or otherwise look up the ASCII code for any given character.

Here is a program to create a valid list of character positions:

```cpp

#include <iostream>
#include <vector>

using namespace std;

int main(void) {
   vector<char> validPAE(128, 0);

   validPAE['b']  = 1; validPAE['A']  = 1; validPAE['B']  = 1;
   validPAE['c']  = 1; validPAE['C']  = 1; validPAE['D']  = 1;
   validPAE['E']  = 1; validPAE['f']  = 1; validPAE['F']  = 1;
   validPAE['g']  = 1; validPAE['G']  = 1; validPAE['i']  = 1;
   validPAE['n']  = 1; validPAE['o']  = 1; validPAE['q']  = 1;
   validPAE['r']  = 1; validPAE['t']  = 1; validPAE['x']  = 1;
   validPAE['%']  = 1; validPAE['-']  = 1; validPAE['+']  = 1;
   validPAE['$']  = 1; validPAE['[']  = 1; validPAE[']']  = 1;
   validPAE['@']  = 1; validPAE[' ']  = 1; validPAE['^']  = 1;
   validPAE['\t'] = 1; validPAE['=']  = 1; validPAE['(']  = 1;
   validPAE[')']  = 1; validPAE['{']  = 1; validPAE['}']  = 1;
   validPAE[';']  = 1; validPAE['!']  = 1; validPAE['/']  = 1;
   validPAE['\''] = 1; validPAE[',']  = 1; validPAE[':']  = 1;
   validPAE['?']  = 1; validPAE['.']  = 1; validPAE['\n'] = 1;
   validPAE['\r'] = 1;

   cout << "vector<char> validPAE(\n";
   for (size_t i = 0; i < validPAE.size(); i++) {
      cout << (int)validPAE[i];
      if (i < 127)       cout << ",";
      if (!((i+1) % 32)) cout << endl;
   }
   cout << ");" << endl;

   return 0;
}
```

Which creates the validation array:

```cpp
vector<char> validPAE(
0,0,0,0,0,0,0,0,0,1,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
1,1,0,0,1,1,0,1,1,1,0,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,1,1,0,1,0,1,
1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,1,0,
0,0,1,1,0,0,1,1,0,1,0,0,0,0,1,1,0,1,1,0,1,0,0,0,1,0,0,1,0,1,0,0
);
```

**Comment by craigsapp** [comment]
There is a question of where to do the validation (include the check for parameters such as key and time signatures), so perhaps @lpugin can add by adjusting the following function (mostly convert `cerr` to a `LogWarning()`:

```cpp

#include <iostream>
#include <iomanip>
#include <vector>


//////////////////////////////
//
// PAEInput::ValidatePAEString -- Returns true if all character in string
//    are valid in PAE; otherwise, returns false.  Also reports any invalid
//    character in a warning message.
//

bool /* PAEInput:: */ ValidatePAEString(const std::string &input) {
   static std::vector<char> validPAE {
      0,0,0,0,0,0,0,0,0,1,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
      1,1,0,0,1,1,0,1,1,1,0,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,1,1,0,1,0,1,
      1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,1,0,
      0,0,1,1,0,0,1,1,0,1,0,0,0,0,1,1,0,1,1,0,1,0,0,0,1,0,0,1,0,1,0,0
   };
   bool status = true;
   for (const char &ch : input) {
      if (ch < 0) {
         std::cerr << "Invalid ASCII hex byte: " << std::hex << (unsigned int)((unsigned char)ch) << std::dec << std::endl;
         status = false;
         continue;
      }
      if (!validPAE[ch]) {
         std::cerr << "Invalid ASCII character: " << ch << std::endl;
         status = false;
      }
   }
   return status;
}


int main(void) {

	// High-bit junk present:
	std::string test = "ABCDEFG√";
	bool status = ValidatePAEString(test);
	std::cerr << "TEST = " << test << "\tSTATUS = " << status << std::endl;

	// Invalid UTF-7 character:
	test = "AaBCDEFG";
	status = ValidatePAEString(test);
	std::cerr << "TEST = " << test << "\tSTATUS = " << status << std::endl;

	// Valid characters:
	test = "ABCDEFG";
	status = ValidatePAEString(test);
	std::cerr << "TEST = " << test << "\tSTATUS = " << status << std::endl;

	return 0;
}
```

(Also I created a static variable which he does not like :-)

Expected output:

```
Invalid ASCII hex byte: e2
Invalid ASCII hex byte: 88
Invalid ASCII hex byte: 9a
TEST = ABCDEFG√	STATUS = 0
Invalid ASCII character: a
TEST = AaBCDEFG	STATUS = 0
TEST = ABCDEFG	STATUS = 1
```

**Comment by rettinghaus** [comment]
I'm wondering why Verovio should do this at all? In a web frontend with an PAE editor doing it directly in JavaScript would make more sense to me. Or is Verovio crashing with nonsense input here?

**Comment by ahankinson** [comment]
The new PAE parser has a validation mode as part of its input process. This is just an enhancement to an existing feature.

**Comment by ahankinson** [comment]
Fixed in 841d9bb81b7108317f9b3cd256748c67b3a5f0a7
